### ISSUE TITLE
UndoRedo documentation lacks information about 800ms interval, and needs clarfication on merging

### ISSUE THREAD

**Issue by TheSofox** [godotengine/godot-docs#8533]
**Your Godot version:** 4.2.rc

**Issue description:** 

UndoRedo documentation is lacking information about how merging of actions work. It only gives a brief overview of MERGE_DISABLE, MERGE_ENDS and MERGE_ALL and doesn't give any code examples on how their behaviour differs. In order to fully understand intended behaviour, I made my own chart:

![image](https://github.com/godotengine/godot-docs/assets/3089525/f21027d2-b059-4ffb-815b-f1302d0491a9)

Also of note is that the documentation makes no mention that actions have to take place within 800ms of eachother in order to be merged. I've seen multiple people being surprised by this behavior, including code contributors.

**URL to the documentation page:** https://docs.godotengine.org/en/stable/classes/class_undoredo.html

**Comment by AnidemDex** [comment]
Note that actions are first `Do` and then `Undo` for `MERGE_DISABLE`

**Comment by AThousandShips** [comment]
I'd argue these are implementation details at least to some degree and aren't guaranteed to be the same, and by documenting them they're made into API details instead that we can't alter if needed just to improve things

**Comment by TheSofox** [comment]
I see your point, but this is a fundamental aspect of how the current UndoRedo system works. I've fixed at least one bug caused by someone who didn't appreciate that the 800ms was even a thing. Likewise, several developers are taken off guard by this.

If the documentation says "If you make three actions of the same name with this setting they'll all be merged together" you can't just have a hidden condition of "Oh, but they have to be less than 800ms apart." It's just a total gotcha.

I mean take a look at this: https://github.com/godotengine/godot/issues/26118 . A merge was causing problems, so they tried to fix the problem by lowering the interval manually, not realising the problem was in the merge itself and lowering the interval just made the merges less frequent so the problem didn't come up as much.

Speaking of altering it in the future, I personally think the 800ms should not be hardcoded and either a Project or Editor setting. I mean, the TextEdit actually has a completely separate Undo system, but its interval is cutomisable in Project Settings (3 sec default). Remember we have to think of this system both for the Godot Editor itself, and whatever game/project people export from the Editor that uses this system (it can of course be used directly from GDScript).

**Comment by AThousandShips** [comment]
I'd say yes, it should be a setting, *or* it should not be specified, as then we can't change it without a good reason even if it improves the behaviour and code

**Comment by TheSofox** [comment]
Alright, so if its a setting, it should probably be a Project Setting since it will need to run in any exported project. So I suggest GUI -> Timers -> Merge Undo Action Time Limit (ms). Int type. Range could 0 upwards, but we could also make -1 a "infinite" time where it won't matter the distance between actions, if all other criteria is met, they merge.  Thoughts?

**Comment by AnidemDex** [comment]
That sounds different from what this issue is about

**Comment by TheSofox** [comment]
If you're trying to say "You should probably make this a separate issue" then please say it. I've put a lot of work and energy into trying to fix this, and it's very annoying to have someone vaguely say that there's something wrong with what I'm doing without offering any suggestions or direction.

**Comment by TypeOverride** [comment]
yeah. I have a similar problem. I do not want this "time interval" MERGE_ALL. As well it causes strange bugs when I click fast. So it merges wrong!? I need something without time. I think the design is bad.
